File: framework/src/datamirai_engine/tools/builtin/filesystem/write_file.py
"""filesystem/write_file — create or overwrite a file."""

from __future__ import annotations

import os
from typing import Any

from datamirai_engine.core.context import ExecutionContext
from datamirai_engine.tools.base import BaseTool, ToolInput, ToolOutput, ToolSpec


class WriteFileTool(BaseTool):
    spec = ToolSpec(
        tool_type="filesystem/write_file",
        version="1.0.0",
        display_name="Write File",
        description="Creates a new file or overwrites an existing one with the provided content.",
        category="filesystem",
        icon="file-plus",
        intents=[
            "crear un archivo nuevo",
            "sobreescribir el contenido de un archivo",
            "guardar texto en un archivo",
        ],
        inputs=[
            ToolInput(name="path", type="string", required=True, description="Absolute or relative path for the file"),
            ToolInput(name="content", type="string", required=True, description="Content to write"),
        ],
        outputs=[
            ToolOutput(name="path", type="string", description="Resolved absolute path"),
            ToolOutput(name="bytes_written", type="number", description="Number of bytes written"),
            ToolOutput(name="created", type="boolean", description="True if the file was newly created"),
        ],
    )

    async def execute(
        self,
        inputs: dict[str, Any],
        config: dict[str, Any],
        context: ExecutionContext | None,
    ) -> dict[str, Any]:
        path = os.path.abspath(inputs["path"])
        content = inputs["content"]
        created = not os.path.exists(path)

        parent = os.path.dirname(path)
        if parent and not os.path.isdir(parent):
            os.makedirs(parent, exist_ok=True)

        with open(path, "w", encoding="utf-8") as f:
            bytes_written = f.write(content)

        return {
            "path": path,
            "bytes_written": bytes_written,
            "created": created,
        }

File: engine/src/tools/builtin/filesystem/write_file.rs
//! WriteFileTool

use super::*;

// ===========================================================================
// WriteFileTool
// ===========================================================================

fs_tool! {
    struct WriteFileTool, factory WriteFileFactory;
    tool_type = "filesystem/write_file",
    name = "Write File",
    description = "Creates a new file or overwrites an existing one with the provided content.",
    category = "filesystem",
    inputs = [
        field("path", FieldType::String, true, "Absolute or relative path for the file"),
        field("content", FieldType::String, true, "Content to write"),
    ],
    outputs = [
        field("path", FieldType::String, true, "Resolved absolute path"),
        field("bytes_written", FieldType::Number, true, "Number of bytes written"),
        field("created", FieldType::Boolean, true, "True if the file was newly created"),
    ],
    config_fields = []
}

#[async_trait]
impl Tool for WriteFileTool {
    async fn execute(
        &self,
        inputs: HashMap<String, Value>,
        _config: &HashMap<String, Value>,
        _context: &dyn ExecutionContext,
    ) -> Result<HashMap<String, Value>, ToolError> {
        let raw_path = inputs.get("path").and_then(|v| v.as_str()).ok_or_else(|| {
            ToolError::ExecutionFailed {
                tool_type: "filesystem/write_file".into(),
                message: "missing required input: path".into(),
            }
        })?;
        let content = inputs
            .get("content")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::ExecutionFailed {
                tool_type: "filesystem/write_file".into(),
                message: "missing required input: content".into(),
            })?;

        let path = PathBuf::from(raw_path);
        let abs_path = if path.is_absolute() {
            path.clone()
        } else {
            std::env::current_dir()
                .unwrap_or_else(|_| PathBuf::from("."))
                .join(&path)
        };

        let created = !abs_path.exists();

        if let Some(parent) = abs_path.parent() {
            if !parent.exists() {
                fs::create_dir_all(parent).map_err(|e| ToolError::ExecutionFailed {
                    tool_type: "filesystem/write_file".into(),
                    message: format!("Cannot create parent directories: {e}"),
                })?;
            }
        }

        let bytes = content.as_bytes();
        fs::write(&abs_path, bytes).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/write_file".into(),
            message: format!("Cannot write file: {e}"),
        })?;

        let mut out = HashMap::new();
        out.insert("path".to_string(), json!(abs_path.display().to_string()));
        out.insert("bytes_written".to_string(), json!(bytes.len()));
        out.insert("created".to_string(), json!(created));
        Ok(out)
    }
}
